README.md
![Logo](https://github.com/lontivero/Open.Nat/raw/gh-pages/images/logos/128.jpg)

Open.NAT
======

Open.NAT is a lightweight and easy-to-use class library to allow port forwarding in NAT devices that support UPNP (Universal Plug & Play) and/or PMP (Port Mapping Protocol).


Goals
-----
NATed computers cannot be reached from outside and this is particularly painful for peer-to-peer or friend-to-friend software.
The main goal is to simplify communication amoung computers behind NAT devices that support UPNP and/or PMP providing a clean
and easy interface to get the external IP address and map ports and helping you to achieve peer-to-peer communication.

How to use?
-----------
With nuget :
> **Install-Package AscensionGameDev.Open.NAT**

Go on the [nuget website](https://www.nuget.org/packages/AscensionGameDev.Open.NAT/) for more information.

Example
--------

The simplest scenario:

```c#
var discoverer = new NatDiscoverer();
var device = await discoverer.DiscoverDeviceAsync();
var ip = await device.GetExternalIPAsync();
Console.WriteLine("The external IP Address is: {0} ", ip);
```

The following piece of code shows a common scenario: It starts the discovery process for a NAT-UPNP device and onces discovered it creates a port mapping. If no device is found before ten seconds, it fails with NatDeviceNotFoundException.


```c#
var discoverer = new NatDiscoverer();
var cts = new CancellationTokenSource(10000);
var device = await discoverer.DiscoverDeviceAsync(PortMapper.Upnp, cts);

await device.CreatePortMapAsync(new Mapping(Protocol.Tcp, 1600, 1700, "The mapping name"));
```


CHANGELOG.md
# Changelog

## v3
### Version 3.0.0
* Re-organized projects
  * Moved Open.Nat to Source/Open.Nat
  * Re-created Open.Nat.UnitTests as Tests/Open.Nat.Tests
* Open.Nat.sln
  * Removed Open.Nat.ConsoleTest (did not delete)
  * Deleted nuget binaries
  * Added editorconfig
* Open.Nat
  * targets only netstandard2.0
  * minor cleanup of dead code
* Open.Nat.Tests
  * targets net6.0
  * broken IPv6 test disabled

## v2

### Version 2.1.0
* Fixes [defect #43](https://github.com/lontivero/Open.NAT/issues/43).
UpnpNatDevice.CreatePortMapAsync does not await retry
* Fixes [defect #45](https://github.com/lontivero/Open.NAT/issues/45).
UpnpNatDevice.GetSpecificMappingAsync returns wrong protocol

### Version 2.0.16
* Supports .NET 3.5
* Fixes [defect #28](https://github.com/lontivero/Open.NAT/issues/28).
GetSpecificMappingAsync under LINKSYS WRT1900AC AC1900 fails with 501

### Version 2.0.11
* Allows the creation of mappings with arbitrary Private IP address.
* Fixes [defect #22](https://github.com/lontivero/Open.NAT/issues/22).
Routers failed with 404 when service control url had a question mark (?) - DD-WRT Linux base router (and others probably) fails with
402-InvalidArgument when index is out of range. - Some routers retuns invalid mapping entries with empty internal client.

* Fixes [defect #24](https://github.com/lontivero/Open.NAT/issues/24).
GetSpecificMappingEntry fails with
402-InvalidArgument in DD-WRT Linux base router when mapping is not found.

### Version 2.0.10
Fixes [defect #20](https://github.com/lontivero/Open.NAT/issues/20). Absolute service control URL path and query miscalculated.

### Version 2.0.9
* Fixes [defect #16](https://github.com/lontivero/Open.NAT/issues/16)

### Version 2.0.8
* Fixes several defects. [#10](https://github.com/lontivero/Open.NAT/issues/10),
[#11](https://github.com/lontivero/Open.NAT/issues/11), [#12](https://github.com/lontivero/Open.NAT/issues/12),
[#13](https://github.com/lontivero/Open.NAT/issues/13) and [#14](https://github.com/lontivero/Open.NAT/issues/12)

### Version 2.0.0
* Thus version breaks backward compatibility with v1.
* Changes the event-based nature of discovery to an asynchronous one.
* Managed port mapping timelife.

## v1

### Version 1.1.0
* Fix for SSDP Location header.
* After this version Open.NAT breaks backward compatibility.

### Version 1.0.19
* Minor changes previous to v2.

### Version 1.0.18
* Discovery timeout raises an event.
* Permanent mappings are created when NAT only supports that kind of mappings.
* Protocol to use in discovery process can be specified.
* Automatic renew port mappings before expiration.
* Add operations timeout after 4 seconds.
* Add parameters validation in Mapping class.
* Fix UnhandledException event was never raised.

### Version 1.0.17
*  Discovery timeout added.
*  Auto release ports opened in the session.
*  Fix NextSearch to use UtcNow (also performance)
*  Fix LocalIP property after add a port.
*  Tracing improvements

### Version 1.0.16
*  Discovery performance and bandwidth improved
*  Tracing improved
*  Cleaner code


Tests/Open.Nat.Tests/Upnp/UpnpMockServer.cs
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Xml.Linq;

namespace Open.Nat.Tests.Upnp;

internal class UpnpMockServer : IDisposable
{
    private readonly HttpListener _listener;
    private readonly ServerConfiguration _cfg;
    public Action<HttpListenerContext> WhenRequestServiceDesc = WhenRequestService;
    public Action<HttpListenerContext> WhenGetExternalIpAddress = ResponseOk;
    public Action<HttpListenerContext> WhenAddPortMapping = ResponseOk;
    public Action<HttpListenerContext> WhenGetGenericPortMappingEntry = ResponseOk;
    public Action<HttpListenerContext> WhenDeletePortMapping = ResponseOk;
    public Func<string> WhenDiscoveryRequest;

    private string HandleDiscoveryRequest()
    {
        return "HTTP/1.1 200 OK\r\n" +
            "Server: Custom/1.0 UPnP/1.0 Proc/Ver\r\n" +
            "EXT:\r\n" +
            $"Location: {_cfg.ServiceUrl}\r\n" +
            "Cache-Control:max-age=1800\r\n" +
            $"ST:urn:schemas-upnp-org:service:{_cfg.ServiceType}\r\n" +
            $"USN:uuid:0000e068-20a0-00e0-20a0-48a802086048::urn:schemas-upnp-org:service:{_cfg.ServiceType}";
    }

    private static void ResponseOk(HttpListenerContext context)
    {
        context.Response.Status(200, "OK");
    }

    private static void WhenRequestService(HttpListenerContext context)
    {
        var responseBytes = File.OpenRead("..\\..\\Responses\\ServiceDescription.txt");
        responseBytes.CopyTo(context.Response.OutputStream);
        context.Response.OutputStream.Flush();

        context.Response.Status(200, "OK");
    }

    public UpnpMockServer() : this(new ServerConfiguration()) { }

    public UpnpMockServer(ServerConfiguration cfg)
    {
        _cfg = cfg;
        _listener = new HttpListener();
        _listener.Prefixes.Add(cfg.Prefix);
        _listener.AuthenticationSchemes = AuthenticationSchemes.Anonymous;
        WhenDiscoveryRequest = HandleDiscoveryRequest;
    }

    public void Start()
    {
        StartAnnouncer();
        StartServer();
    }

    private void StartAnnouncer()
    {
        Task.Run(
            () =>
            {
                UdpClient udpClient = default!;
                try
                {
                    var remoteIPEndPoint = new IPEndPoint(IPAddress.Any, 0);
                    using (udpClient = new UdpClient(1900))
                    {
                        while (true)
                        {
                            var bytes = udpClient.Receive(ref remoteIPEndPoint);
                            if (bytes == null || bytes.Length == 0) return;

                            var response = WhenDiscoveryRequest();

                            var responseBytes = Encoding.UTF8.GetBytes(response);
                            udpClient.Send(responseBytes, responseBytes.Length, remoteIPEndPoint);
                        }
                    }
                }
                catch (Exception exception)
                {
                    Console.Error.WriteLine(exception);
                }
                finally
                {
                    udpClient?.Close();
                    udpClient?.Dispose();
                }
            });
    }

    private void StartServer()
    {
        _listener.Start();
        Task.Run(() =>
        {
            while (_listener.IsListening)
            {
                ProcessRequest();
            }
        });
    }

    private void ProcessRequest()
    {
        var result = _listener.BeginGetContext(ListenerCallback, _listener);
        result.AsyncWaitHandle.WaitOne();
    }

    private void ListenerCallback(IAsyncResult result)
    {
        if (!_listener.IsListening) return;
        var context = _listener.EndGetContext(result);
        var request = context.Request;
        if (request.Url.AbsoluteUri == _cfg.ServiceUrl)
        {
            WhenRequestServiceDesc(context);
            return;
        }

        if (request.Url.AbsoluteUri == _cfg.ControlUrl)
        {
            var soapActionHeader = request.Headers["SOAPACTION"];
            soapActionHeader = soapActionHeader.Substring(1, soapActionHeader.Length - 2);

            var soapActionHeaderParts = soapActionHeader.Split(new[] { '#' });
            var serviceType = soapActionHeaderParts[0];
            var soapAction = soapActionHeaderParts[1];
            var buffer = new byte[request.ContentLength64 - 4];
            request.InputStream.Read(buffer, 0, buffer.Length);
            var body = Encoding.UTF8.GetString(buffer);
            var envelop = XElement.Parse(body);

            switch (soapAction)
            {
                case "GetExternalIPAddress":
                    WhenGetExternalIpAddress(context);
                    return;
                case "AddPortMapping":
                    WhenAddPortMapping(context);
                    return;
                case "GetGenericPortMappingEntry":
                    WhenGetGenericPortMappingEntry(context);
                    return;
                case "DeletePortMapping":
                    WhenDeletePortMapping(context);
                    return;
            }
            context.Response.Status(200, "OK");
            return;
        }
        context.Response.Status(500, "Internal Server Error");
    }

    public void Dispose()
    {
        _listener.Close();
    }
}

internal static class HttpListenerResponseExtensions
{
    public static void Status(this HttpListenerResponse res, int statusCode, string description)
    {
        res.StatusCode = statusCode;
        res.StatusDescription = description;
        res.Close();
    }
}

internal class ServerConfiguration
{
    public ServerConfiguration()
    {
        ServiceType = "WANIPConnection:1";
        Prefix = "http://127.0.0.1:5431/";
        ServiceUrl = "/dyndev/uuid:0000e068-20a0-00e0-20a0-48a8000808e0";
        ControlUrl = "/uuid:0000e068-20a0-00e0-20a0-48a802086048/" + ServiceType;
    }

    public string ControlUrl { get; set; }

    public string ServiceUrl { get; set; }

    public string ServiceType { get; set; }

    public string Prefix { get; set; }
}


Tests/Open.Nat.Tests/Upnp/UpnpNatDeviceInfoTests.cs
using System.Net;

using NUnit.Framework;

namespace Open.Nat.Tests.Upnp;

public class UpnpNatDeviceInfoTests
{
    [SetUp]
    public void Setup()
    {
    }

    [Test]
    public void TestServiceControlUri()
    {
        var info = new UpnpNatDeviceInfo(IPAddress.Loopback, new Uri("http://127.0.0.1:3221"), "/control?WANIPConnection", null);
        Assert.AreEqual("http://127.0.0.1:3221/control?WANIPConnection", info.ServiceControlUri.ToString());
    }
}


Tests/Open.Nat.Tests/Upnp/UpnpNatDeviceIpv6Tests.cs
using System.Net;
using System.Text;

using NUnit.Framework;

namespace Open.Nat.Tests.Upnp;
internal class UpnpNatDeviceIpv6Tests
{
    private UpnpMockServer _server = default!;
    private ServerConfiguration _cfg = default!;

    [SetUp]
    public void Setup()
    {
        _cfg = new ServerConfiguration
        {
            Prefix = "http://[::1]:5431/",
            ServiceUrl = "http://[::1]:5431/dyndev/uuid:0000e068-20a0-00e0-20a0-48a8000808e0",
            ControlUrl = "http://[::1]:5431/uuid:0000e068-20a0-00e0-20a0-48a802086048/WANIPConnection:1"
        };
        _server = new UpnpMockServer(_cfg);
        _server.Start();
    }

    [TearDown]
    public void TearDown()
    {
        _server.Dispose();
    }

    [Ignore("Currently returns external IPv4 address, never hits the mock request handlers specified in the test.")]
    [Test]
    public async Task Connect()
    {
        _server.WhenDiscoveryRequest = () =>
                  "HTTP/1.1 200 OK\r\n"
                + "Server: Custom/1.0 UPnP/1.0 Proc/Ver\r\n"
                + "EXT:\r\n"
                + "Location: http://[::1]:5431/dyndev/uuid:0000e068-20a0-00e0-20a0-48a8000808e0\r\n"
                + "Cache-Control:max-age=1800\r\n"
                + "ST:urn:schemas-upnp-org:service:WANIPConnection:1\r\n"
                + "USN:uuid:0000e068-20a0-00e0-20a0-48a802086048::urn:schemas-upnp-org:service:WANIPConnection:1";

        _server.WhenGetExternalIpAddress = (ctx) =>
        {
            var responseXml = "<?xml version=\"1.0\"?>" +
                "<s:Envelope xmlns:s=\"http://schemas.xmlsoap.org/soap/envelope/\" " +
                "s:encodingStyle=\"http://schemas.xmlsoap.org/soap/encoding/\">" +
                "<s:Body>" +
                "<m:GetExternalIPAddressResponse xmlns:m=\"urn:schemas-upnp-org:service:WANIPConnection:1\">" +
                "<NewExternalIPAddress>FE80::0202:B3FF:FE1E:8329</NewExternalIPAddress>" +
                "</m:GetExternalIPAddressResponse>" +
                "</s:Body>" +
                "</s:Envelope>";
            var bytes = Encoding.UTF8.GetBytes(responseXml);
            var response = ctx.Response;
            response.OutputStream.Write(bytes, 0, bytes.Length);
            response.OutputStream.Flush();
            response.StatusCode = 200;
            response.StatusDescription = "OK";
            response.Close();
        };

        var nat = new NatDiscoverer();
        var cts = new CancellationTokenSource(5000);
        var device = await nat.DiscoverDeviceAsync(PortMapper.Upnp, cts);
        Assert.IsNotNull(device);

        var ip = await device.GetExternalIPAsync();
        Assert.AreEqual(IPAddress.Parse("FE80::0202:B3FF:FE1E:8329"), ip);
    }
}


Source/Open.Nat/NatDiscoverer.cs
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace Open.Nat
{
    /// <summary>
    /// 
    /// </summary>
    public class NatDiscoverer
    {
        /// <summary>
        /// The <see cref="http://msdn.microsoft.com/en-us/library/vstudio/system.diagnostics.tracesource">TraceSource</see> instance
        /// used for debugging and <see cref="https://github.com/lontivero/Open.Nat/wiki/Troubleshooting">Troubleshooting</see>
        /// </summary>
        /// <example>
        /// NatUtility.TraceSource.Switch.Level = SourceLevels.Verbose;
        /// NatUtility.TraceSource.Listeners.Add(new ConsoleListener());
        /// </example>
        /// <remarks>
        /// At least one trace listener has to be added to the Listeners collection if a trace is required; if no listener is added
        /// there will no be tracing to analyse.
        /// </remarks>
        /// <remarks>
        /// Open.NAT only supports SourceLevels.Verbose, SourceLevels.Error, SourceLevels.Warning and SourceLevels.Information.
        /// </remarks>
        public readonly static TraceSource TraceSource = new TraceSource("Open.NAT");

        private static readonly Dictionary<string, NatDevice> Devices = new Dictionary<string, NatDevice>();

        // Finalizer is never used however its destructor, that releases the open ports, is invoked by the
        // process as part of the shuting down step. So, don't remove it!
        private static readonly Finalizer Finalizer = new Finalizer();
        internal static readonly Timer RenewTimer = new Timer(RenewMappings, null, 5000, 2000);

        /// <summary>
        /// Discovers and returns an UPnp or Pmp NAT device; otherwise a <see cref="NatDeviceNotFoundException">NatDeviceNotFoundException</see>
        /// exception is thrown after 3 seconds. 
        /// </summary>
        /// <returns>A NAT device</returns>
        /// <exception cref="NatDeviceNotFoundException">when no NAT found before 3 seconds.</exception>
        public async Task<NatDevice> DiscoverDeviceAsync()
        {
            var cts = new CancellationTokenSource(3 * 1000);
            return await DiscoverDeviceAsync(PortMapper.Pmp | PortMapper.Upnp, cts);
        }

        /// <summary>
        /// Discovers and returns a NAT device for the specified type; otherwise a <see cref="NatDeviceNotFoundException">NatDeviceNotFoundException</see> 
        /// exception is thrown when it is cancelled. 
        /// </summary>
        /// <remarks>
        /// It allows to specify the NAT type to discover as well as the cancellation token in order.
        /// </remarks>
        /// <param name="portMapper">Port mapper protocol; Upnp, Pmp or both</param>
        /// <param name="cancellationTokenSource">Cancellation token source for cancelling the discovery process</param>
        /// <returns>A NAT device</returns>
        /// <exception cref="NatDeviceNotFoundException">when no NAT found before cancellation</exception>
        public async Task<NatDevice> DiscoverDeviceAsync(PortMapper portMapper, CancellationTokenSource cancellationTokenSource)
        {
            Guard.IsTrue(portMapper.HasFlag(PortMapper.Upnp) || portMapper.HasFlag(PortMapper.Pmp), "portMapper");
            Guard.IsNotNull(cancellationTokenSource, "cancellationTokenSource");

            var devices = await DiscoverAsync(portMapper, true, cancellationTokenSource);
            var deviceList = devices.ToList();
            var device = deviceList.FirstOrDefault();
            if (device == null)
            {
                TraceSource.LogInfo("Device not found. Common reasons:");
                TraceSource.LogInfo("\t* No device is present or,");
                TraceSource.LogInfo("\t* Upnp is disabled in the router or");
                TraceSource.LogInfo("\t* Antivirus software is filtering SSDP (discovery protocol).");
                throw new NatDeviceNotFoundException();
            }
            return device;
        }

        /// <summary>
        /// Discovers and returns all NAT devices for the specified type. If no NAT device is found it returns an empty enumerable
        /// </summary>
        /// <param name="portMapper">Port mapper protocol; Upnp, Pmp or both</param>
        /// <param name="cancellationTokenSource">Cancellation token source for cancelling the discovery process</param>
        /// <returns>All found NAT devices</returns>
        public async Task<IEnumerable<NatDevice>> DiscoverDevicesAsync(PortMapper portMapper, CancellationTokenSource cancellationTokenSource)
        {
            Guard.IsTrue(portMapper.HasFlag(PortMapper.Upnp) || portMapper.HasFlag(PortMapper.Pmp), "portMapper");
            Guard.IsNotNull(cancellationTokenSource, "cancellationTokenSource");

            var devices = await DiscoverAsync(portMapper, false, cancellationTokenSource);
            return devices.ToArray();
        }

        private async Task<IEnumerable<NatDevice>> DiscoverAsync(PortMapper portMapper, bool onlyOne, CancellationTokenSource cts)
        {
            TraceSource.LogInfo("Start Discovery");
            var searcherTasks = new List<Task<IEnumerable<NatDevice>>>();
            if (portMapper.HasFlag(PortMapper.Upnp))
            {
                var upnpSearcher = new UpnpSearcher(new IPAddressesProvider());
                upnpSearcher.DeviceFound += (sender, args) => { if (onlyOne) cts.Cancel(); };
                searcherTasks.Add(upnpSearcher.Search(cts.Token));
            }
            if (portMapper.HasFlag(PortMapper.Pmp))
            {
                var pmpSearcher = new PmpSearcher(new IPAddressesProvider());
                pmpSearcher.DeviceFound += (sender, args) => { if (onlyOne) cts.Cancel(); };
                searcherTasks.Add(pmpSearcher.Search(cts.Token));
            }

            await Task.WhenAll(searcherTasks);
            TraceSource.LogInfo("Stop Discovery");

            var devices = searcherTasks.SelectMany(x => x.Result);
            foreach (var device in devices)
            {
                var key = device.ToString();
                NatDevice nat;
                if (Devices.TryGetValue(key, out nat))
                {
                    nat.Touch();
                }
                else
                {
                    Devices.Add(key, device);
                }
            }
            return devices;
        }

        /// <summary>
        /// Release all ports opened by Open.NAT. 
        /// </summary>
        /// <remarks>
        /// If ReleaseOnShutdown value is true, it release all the mappings created through the library.
        /// </remarks>
        public static void ReleaseAll()
        {
            foreach (var device in Devices.Values)
            {
                device.ReleaseAll();
            }
        }

        internal static void ReleaseSessionMappings()
        {
            foreach (var device in Devices.Values)
            {
                device.ReleaseSessionMappings();
            }
        }

        private static void RenewMappings(object state)
        {
            Task.Factory.StartNew(async () =>
            {
                foreach (var device in Devices.Values)
                {
                    await device.RenewMappings();
                }
            });
        }
    }
}


Source/Open.Nat/Upnp/UpnpNatDeviceInfo.cs
//
// Authors:
//   Lucas Ontivero lucasontivero@gmail.com 
//
// Copyright (C) 2014 Lucas Ontivero
//
// Permission is hereby granted, free of charge, to any person obtaining
// a copy of this software and associated documentation files (the
// "Software"), to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to
// permit persons to whom the Software is furnished to do so, subject to
// the following conditions:
// 
// The above copyright notice and this permission notice shall be
// included in all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

using System;
using System.Net;

namespace Open.Nat
{
    internal class UpnpNatDeviceInfo
    {
        public UpnpNatDeviceInfo(IPAddress localAddress, Uri locationUri, string serviceControlUrl, string serviceType)
        {
            LocalAddress = localAddress;
            ServiceType = serviceType;
            HostEndPoint = new IPEndPoint(IPAddress.Parse(locationUri.Host), locationUri.Port);

            if (Uri.IsWellFormedUriString(serviceControlUrl, UriKind.Absolute))
            {
                var serviceControlUri = new Uri(serviceControlUrl);
                IPEndPoint old = HostEndPoint;
                serviceControlUrl = serviceControlUri.PathAndQuery;

                NatDiscoverer.TraceSource.LogInfo($"{old}: Absolute URI detected. Host address is now: {HostEndPoint}");
                NatDiscoverer.TraceSource.LogInfo($"{HostEndPoint}: New control url: {serviceControlUrl}");
            }

            var builder = new UriBuilder("http", locationUri.Host, locationUri.Port);
            ServiceControlUri = new Uri(builder.Uri, serviceControlUrl);
        }

        public IPEndPoint HostEndPoint { get; private set; }

        public IPAddress LocalAddress { get; private set; }

        public Uri ServiceControlUri { get; private set; }

        public string ServiceType { get; private set; }
    }
}


Source/Open.Nat/Upnp/UpnpNatDevice.cs
//
// Authors:
//   Alan McGovern alan.mcgovern@gmail.com
//   Ben Motmans <ben.motmans@gmail.com>
//   Lucas Ontivero lucasontivero@gmail.com 
//
// Copyright (C) 2006 Alan McGovern
// Copyright (C) 2007 Ben Motmans
// Copyright (C) 2014 Lucas Ontivero
//
// Permission is hereby granted, free of charge, to any person obtaining
// a copy of this software and associated documentation files (the
// "Software"), to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to
// permit persons to whom the Software is furnished to do so, subject to
// the following conditions:
// 
// The above copyright notice and this permission notice shall be
// included in all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

using System;
using System.Collections;
using System.Collections.Generic;
using System.Net;
using System.Threading.Tasks;

namespace Open.Nat
{
    internal sealed class UpnpNatDevice : NatDevice
    {
        internal readonly UpnpNatDeviceInfo DeviceInfo;
        private readonly SoapClient _soapClient;

        internal UpnpNatDevice(UpnpNatDeviceInfo deviceInfo)
        {
            Touch();
            DeviceInfo = deviceInfo;
            _soapClient = new SoapClient(DeviceInfo.ServiceControlUri, DeviceInfo.ServiceType);
        }

        public override async Task<IPAddress> GetExternalIPAsync()
        {
            NatDiscoverer.TraceSource.LogInfo("GetExternalIPAsync - Getting external IP address");
            var message = new GetExternalIPAddressRequestMessage();
            var responseData = await _soapClient
                .InvokeAsync("GetExternalIPAddress", message.ToXml())
                .TimeoutAfter(TimeSpan.FromSeconds(4));

            var response = new GetExternalIPAddressResponseMessage(responseData, DeviceInfo.ServiceType);
            return response.ExternalIPAddress;
        }

        public override async Task CreatePortMapAsync(Mapping mapping)
        {
            Guard.IsNotNull(mapping, "mapping");
            if (mapping.PrivateIP.Equals(IPAddress.None)) mapping.PrivateIP = DeviceInfo.LocalAddress;

            NatDiscoverer.TraceSource.LogInfo("CreatePortMapAsync - Creating port mapping {0}", mapping);
            bool retry = false;
            try
            {
                var message = new CreatePortMappingRequestMessage(mapping);
                await _soapClient
                    .InvokeAsync("AddPortMapping", message.ToXml())
                    .TimeoutAfter(TimeSpan.FromSeconds(4));
                RegisterMapping(mapping);
            }
            catch (MappingException me)
            {
                switch (me.ErrorCode)
                {
                    case UpnpConstants.OnlyPermanentLeasesSupported:
                        NatDiscoverer.TraceSource.LogWarn("Only Permanent Leases Supported - There is no warranty it will be closed");
                        mapping.Lifetime = 0;
                        // We create the mapping anyway. It must be released on shutdown.
                        mapping.LifetimeType = MappingLifetime.ForcedSession;
                        retry = true;
                        break;
                    case UpnpConstants.SamePortValuesRequired:
                        NatDiscoverer.TraceSource.LogWarn("Same Port Values Required - Using internal port {0}", mapping.PrivatePort);
                        mapping.PublicPort = mapping.PrivatePort;
                        retry = true;
                        break;
                    case UpnpConstants.RemoteHostOnlySupportsWildcard:
                        NatDiscoverer.TraceSource.LogWarn("Remote Host Only Supports Wildcard");
                        mapping.PublicIP = IPAddress.None;
                        retry = true;
                        break;
                    case UpnpConstants.ExternalPortOnlySupportsWildcard:
                        NatDiscoverer.TraceSource.LogWarn("External Port Only Supports Wildcard");
                        throw;
                    case UpnpConstants.ConflictInMappingEntry:
                        NatDiscoverer.TraceSource.LogWarn("Conflict with an already existing mapping");
                        throw;

                    default:
                        throw;
                }
            }
            if (retry)
                await CreatePortMapAsync(mapping);
        }

        public override async Task DeletePortMapAsync(Mapping mapping)
        {
            Guard.IsNotNull(mapping, "mapping");

            if (mapping.PrivateIP.Equals(IPAddress.None)) mapping.PrivateIP = DeviceInfo.LocalAddress;

            NatDiscoverer.TraceSource.LogInfo("DeletePortMapAsync - Deleteing port mapping {0}", mapping);

            try
            {
                var message = new DeletePortMappingRequestMessage(mapping);
                await _soapClient
                    .InvokeAsync("DeletePortMapping", message.ToXml())
                    .TimeoutAfter(TimeSpan.FromSeconds(4));
                UnregisterMapping(mapping);
            }
            catch (MappingException e)
            {
                if (e.ErrorCode != UpnpConstants.NoSuchEntryInArray) throw;
            }
        }

        public override async Task<IEnumerable<Mapping>> GetAllMappingsAsync()
        {
            var index = 0;
            var mappings = new List<Mapping>();

            NatDiscoverer.TraceSource.LogInfo("GetAllMappingsAsync - Getting all mappings");
            while (true)
            {
                try
                {
                    var message = new GetGenericPortMappingEntry(index++);

                    var responseData = await _soapClient
                        .InvokeAsync("GetGenericPortMappingEntry", message.ToXml())
                        .TimeoutAfter(TimeSpan.FromSeconds(4));

                    var responseMessage = new GetPortMappingEntryResponseMessage(responseData, DeviceInfo.ServiceType, true);

                    IPAddress internalClientIp;
                    if (!IPAddress.TryParse(responseMessage.InternalClient, out internalClientIp))
                    {
                        NatDiscoverer.TraceSource.LogWarn("InternalClient is not an IP address. Mapping ignored!");
                        continue;
                    }

                    var mapping = new Mapping(responseMessage.Protocol
                        , internalClientIp
                        , responseMessage.InternalPort
                        , responseMessage.ExternalPort
                        , responseMessage.LeaseDuration
                        , responseMessage.PortMappingDescription);
                    mappings.Add(mapping);
                }
                catch (MappingException e)
                {
                    // there are no more mappings
                    if (e.ErrorCode == UpnpConstants.SpecifiedArrayIndexInvalid
                     || e.ErrorCode == UpnpConstants.NoSuchEntryInArray
                     // DD-WRT Linux base router (and others probably) fails with 402-InvalidArgument when index is out of range
                     || e.ErrorCode == UpnpConstants.InvalidArguments
                     // LINKSYS WRT1900AC AC1900 it returns errocode 501-PAL_UPNP_SOAP_E_ACTION_FAILED
                     || e.ErrorCode == UpnpConstants.ActionFailed)
                    {
                        NatDiscoverer.TraceSource.LogWarn("Router failed with {0}-{1}. No more mappings is assumed.", e.ErrorCode, e.ErrorText);
                        break;
                    }
                    throw;
                }
            }

            return mappings.ToArray();
        }

        public override async Task<Mapping> GetSpecificMappingAsync(Protocol protocol, int publicPort)
        {
            Guard.IsTrue(protocol == Protocol.Tcp || protocol == Protocol.Udp, "protocol");
            Guard.IsInRange(publicPort, 0, ushort.MaxValue, "port");

            NatDiscoverer.TraceSource.LogInfo("GetSpecificMappingAsync - Getting mapping for protocol: {0} port: {1}", Enum.GetName(typeof(Protocol), protocol), publicPort);

            try
            {
                var message = new GetSpecificPortMappingEntryRequestMessage(protocol, publicPort);
                var responseData = await _soapClient
                    .InvokeAsync("GetSpecificPortMappingEntry", message.ToXml())
                    .TimeoutAfter(TimeSpan.FromSeconds(4));

                var messageResponse = new GetPortMappingEntryResponseMessage(responseData, DeviceInfo.ServiceType, false);

                if (messageResponse.Protocol != protocol)
                    NatDiscoverer.TraceSource.LogWarn("Router responded to a protocol {0} query with a protocol {1} answer, work around applied.", protocol, messageResponse.Protocol);

                return new Mapping(protocol
                    , IPAddress.Parse(messageResponse.InternalClient)
                    , messageResponse.InternalPort
                    , publicPort // messageResponse.ExternalPort is short.MaxValue
                    , messageResponse.LeaseDuration
                    , messageResponse.PortMappingDescription);
            }
            catch (MappingException e)
            {
                // there are no more mappings
                if (e.ErrorCode == UpnpConstants.SpecifiedArrayIndexInvalid
                 || e.ErrorCode == UpnpConstants.NoSuchEntryInArray
                 // DD-WRT Linux base router (and others probably) fails with 402-InvalidArgument when index is out of range
                 || e.ErrorCode == UpnpConstants.InvalidArguments
                 // LINKSYS WRT1900AC AC1900 it returns errocode 501-PAL_UPNP_SOAP_E_ACTION_FAILED
                 || e.ErrorCode == UpnpConstants.ActionFailed)
                {
                    NatDiscoverer.TraceSource.LogWarn("Router failed with {0}-{1}. No more mappings is assumed.", e.ErrorCode, e.ErrorText);
                    return null;
                }
                throw;
            }
        }

        public override string ToString()
        {
            //GetExternalIP is blocking and can throw exceptions, can't use it here.
            return String.Format(
                "EndPoint: {0}\nControl Url: {1}\nService Type: {2}\nLast Seen: {3}",
                DeviceInfo.HostEndPoint, DeviceInfo.ServiceControlUri, DeviceInfo.ServiceType, LastSeen);
        }
    }
}
